### PR TITLE
TSP implementation

### PR DESCRIPTION
**Pull request by BastienFaivre**
I implemented the TSP algorithm

### CODE REVIEW

File: algorithms/dp/bitmasking_tsp.py [hunk 1]
```diff
@@ -1,15 +1,187 @@
...
+    
+    # store the result in the DP matrix
+    DP[index][mask] = minDist
+
+    return DP[index][mask]
+
 """
 Finds the length of the shortest Euler circuit using bitmasking and dynamic programming (DP).
 :param nodes: Contains integer tuples of the node coordinates.
 :returns: the length of the shortest Euler circuit.
```
Comment (AmandaStromdahl) [inline algorithms/dp/bitmasking_tsp.py:166]: Could you update the comment by including the other two params and making sure the description of _nodes_ is correct?

File: algorithms/dp/bitmasking_tsp.py [hunk 2]
```diff
@@ -1,15 +1,187 @@
...
+
+    return dist
+
+"""
+Initialize the variables
+
+:param nodes: The nodes given
+:param nbRow: The number of rows
+:param nbColumn: The number of columns
+"""
```
Comment (AmandaStromdahl) [inline algorithms/dp/bitmasking_tsp.py:102]: Could you add a description of the return statement to the comment?

File: algorithms/dp/bitmasking_tsp.py (hunk 1 above)
Comment (BastienFaivre) [reply algorithms/dp/bitmasking_tsp.py:166]: Sure, I forgot to update it, good catch !

File: algorithms/dp/bitmasking_tsp.py (hunk 2 above)
Comment (BastienFaivre) [reply algorithms/dp/bitmasking_tsp.py:102]: I'll do it, thank you for catching this miss

File: algorithms/dp/bitmasking_tsp.py [hunk 3]
```diff
@@ -1,15 +1,187 @@
...
+            if node != '.' and node != '*' and node != '#':
+                raise ValueError(f'The given node "{node}" in row {row} is not valid !')
+
+"""
+Check if the given position is valid
+
+:param nodes: The nodes
+:param nbRow: The number of rows
+:param nbColumn: The number of columns
+:returns: True if the position is valid, False otherwise
```
Comment (AmandaStromdahl) [inline algorithms/dp/bitmasking_tsp.py:37]: Could you add descriptions of _r_ and _c_ as well?

File: algorithms/dp/bitmasking_tsp.py [hunk 4]
```diff
@@ -1,15 +1,187 @@
...
+
+"""
+Check that the given nodes is valid
+
+:param nodes: The nodes to check
+:param nbRow: The number of rows
+:param nbColumn: The number of columns
+:returns: Nothing but raises error if invalid nodes
+"""
+def check_argument(nodes: list, nbRow: int, nbColumn: int):
```
Comment (AmandaStromdahl) [inline algorithms/dp/bitmasking_tsp.py:19]: I think it would be more logical to place _check_argument_ right above _initialization_.

**Comment by AmandaStromdahl** [comment]
Great work! As far as I can tell, the implementation seems correct. I had a few minor comments about the descriptions of the functions, but other than that I'm fine with merging this branch. After that we just need to update some of the requirements and the tests.

File: algorithms/dp/bitmasking_tsp.py [hunk 5]
```diff
@@ -1,15 +1,187 @@
...
+        raise ValueError(f'The number of rows in nodes ({len(nodes)}) does not match the given number of row: {nbRow} !')
+    for row in nodes:
+        if not isinstance(row, list):
+            raise ValueError(f'The given row {row} is not valid !')
+        if len(row) != nbColumn:
+            raise ValueError(f'The number of columns in row {row} does not match the given number of column: {nbColumn} !')
+        for node in row:
+            if node != '.' and node != '*' and node != '#':
+                raise ValueError(f'The given node "{node}" in row {row} is not valid !')
+
```
Comment (AmandaStromdahl) [inline algorithms/dp/bitmasking_tsp.py:30]: Could you also make sure that _check_argument_ verifies that _nodes_ isn't empty? I ran some tests on this code on a local branch and found that all tests passed except for the test in which nodes=[].

File: algorithms/dp/bitmasking_tsp.py (hunk 4 above)
Comment (BastienFaivre) [reply algorithms/dp/bitmasking_tsp.py:19]: I agree with this change

File: algorithms/dp/bitmasking_tsp.py (hunk 5 above)
Comment (BastienFaivre) [reply algorithms/dp/bitmasking_tsp.py:30]: Sure, I'll add this special case when checking the argument

**Comment by AmandaStromdahl** [comment]
The new changes look good. I'll merge this branch.

**Pull request merged**